gtk4/src/subclass/entry_buffer.rs

```diff
@@ -0,0 +1,284 @@
+        klass.delete_text = Some(entry_buffer_delete_text::<T>);
+        klass.deleted_text = Some(entry_buffer_deleted_text::<T>);
+        klass.get_length = Some(entry_buffer_get_length::<T>);
+        klass.get_text = Some(entry_buffer_get_text::<T>);
+        klass.insert_text = Some(entry_buffer_insert_text::<T>);
+        klass.inserted_text = Some(entry_buffer_inserted_text::<T>);
+    }
+}
+
+unsafe extern "C" fn entry_buffer_delete_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    n_chars: u32,
+) -> u32 {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    let n_chars = if n_chars == u32::MAX {
+        None
+    } else {
+        Some(n_chars)
+    };
+
+    imp.delete_text(wrap.unsafe_cast_ref(), position, n_chars)
+}
+
+unsafe extern "C" fn entry_buffer_deleted_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    n_chars: u32,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    let n_chars = if n_chars == u32::MAX {
+        None
+    } else {
+        Some(n_chars)
+    };
+
+    imp.deleted_text(wrap.unsafe_cast_ref(), position, n_chars)
+}
+
+static GET_TEXT_QUARK: Lazy<glib::Quark> =
+    Lazy::new(|| glib::Quark::from_string("gtk4-rs-subclass-entry-buffer-text"));
+
+unsafe extern "C" fn entry_buffer_get_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    n_bytes: *mut usize,
+) -> *const libc::c_char {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    let ret = imp.get_text(wrap.unsafe_cast_ref());
+    *n_bytes = ret.len();
+    // Ensures that the returned text stays alive for as long as
+    // the entry buffer instance
+    let fullptr = ret.to_glib_full();
+    wrap.set_qdata(*GET_TEXT_QUARK, fullptr);
+    fullptr
+}
+
+unsafe extern "C" fn entry_buffer_get_length<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+) -> u32 {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    imp.get_length(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn entry_buffer_insert_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    charsptr: *const libc::c_char,
+    n_chars: u32,
+) -> u32 {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+    let text: Borrowed<GString> = from_glib_borrow(charsptr);
+
+    let mut char_indices = text.char_indices().map(|(i, _char)| i);
+    let pos_end = if n_chars != u32::MAX {
+        char_indices.nth(n_chars as usize)
+    } else {
+        char_indices.last()
```

Comment:
This gives you the start byte of the last character. What you actually want here is the byte after the end of the last character, or otherwise your `text[0..pos_end]` below cuts the last character

Reply:
And the other case probably also needs a `+1`

Reply:
Now you increase the index in bytes, not characters :) So you'd end up in the middle of a character if you're unlucky.

Reply:
I have did another attempt at it. I think it's probably correct :)